packages/instruction-plans/src/transaction-plan-executor.ts

```diff
@@ -1,7 +1,228 @@
+
+        const cancelHandler = () => {
+            context.canceled = true;
+        };
+        abortSignal?.addEventListener('abort', cancelHandler);
+        const transactionPlanResult = await traverse(plan, context);
+        abortSignal?.removeEventListener('abort', cancelHandler);
+
+        if (context.canceled) {
+            const abortReason = abortSignal?.aborted ? abortSignal.reason : undefined;
+            throw new SolanaError(SOLANA_ERROR__INSTRUCTION_PLANS__FAILED_TO_EXECUTE_TRANSACTION_PLAN, {
+                cause: findErrorFromTransactionPlanResult(transactionPlanResult) ?? abortReason,
+                transactionPlanResult,
+            });
+        }
+
+        return transactionPlanResult;
+    };
+}
+
+type TraverseContext = TransactionPlanExecutorConfig & {
+    abortSignal?: AbortSignal;
+    canceled: boolean;
+};
+
+async function traverse(transactionPlan: TransactionPlan, context: TraverseContext): Promise<TransactionPlanResult> {
+    const kind = transactionPlan.kind;
+    switch (kind) {
+        case 'sequential':
+            return await traverseSequential(transactionPlan, context);
+        case 'parallel':
+            return await traverseParallel(transactionPlan, context);
+        case 'single':
+            return await traverseSingle(transactionPlan, context);
+        default:
+            transactionPlan satisfies never;
+            throw new SolanaError(SOLANA_ERROR__INVARIANT_VIOLATION__INVALID_TRANSACTION_PLAN_KIND, { kind });
+    }
+}
+
+async function traverseSequential(
+    transactionPlan: SequentialTransactionPlan,
+    context: TraverseContext,
+): Promise<TransactionPlanResult> {
+    const results: TransactionPlanResult[] = [];
+
+    for (const subPlan of transactionPlan.plans) {
+        const result = await traverse(subPlan, context);
+        results.push(result);
+    }
+
+    return transactionPlan.divisible
+        ? sequentialTransactionPlanResult(results)
+        : nonDivisibleSequentialTransactionPlanResult(results);
+}
+
+async function traverseParallel(
+    transactionPlan: ParallelTransactionPlan,
+    context: TraverseContext,
+): Promise<TransactionPlanResult> {
+    const results = await concurrentOperations(
+        transactionPlan.plans.map(plan => () => traverse(plan, context)),
+        context.maxConcurrentOperations,
+    );
+
+    return parallelTransactionPlanResult(results);
+}
+
+async function concurrentOperations<T>(operations: (() => Promise<T>)[], maxConcurrentOperations?: number) {
+    const results: T[] = [];
+    const thread = {
+        fail: (_e: unknown) => {},
+        unblock: () => {},
+    };
+
+    function okToContinue() {
+        return new Promise<void>((resolve, reject) => {
+            thread.fail = reject;
+            thread.unblock = resolve;
+        });
+    }
+
+    let numInFlightOperations = 0;
+    for (let index = 0; index < operations.length; index++) {
+        if (numInFlightOperations === maxConcurrentOperations) {
```

Comment:
We should probably throw an invariant error if `maxConcurrentOperations` is `<1`. If someone sets it to zero the program will just halt.